module/zfs/vdev.c

```diff
@@ -2689,42 +2689,79 @@ vdev_accessible(vdev_t *vd, zio_t *zio)
  * Get statistics for the given vdev.
  */
 void
-vdev_get_stats(vdev_t *vd, vdev_stat_t *vs)
+__vdev_get_stats(vdev_t *vd, vdev_stat_t *vs)
 {
-	spa_t *spa = vd->vdev_spa;
-	vdev_t *rvd = spa->spa_root_vdev;
-	int c, t;
-
-	ASSERT(spa_config_held(spa, SCL_ALL, RW_READER) != 0);
-
-	mutex_enter(&vd->vdev_stat_lock);
-	bcopy(&vd->vdev_stat, vs, sizeof (*vs));
-	vs->vs_timestamp = gethrtime() - vs->vs_timestamp;
-	vs->vs_state = vd->vdev_state;
-	vs->vs_rsize = vdev_get_min_asize(vd);
-	if (vd->vdev_ops->vdev_op_leaf)
-		vs->vs_rsize += VDEV_LABEL_START_SIZE + VDEV_LABEL_END_SIZE;
-	vs->vs_esize = vd->vdev_max_asize - vd->vdev_asize;
-	if (vd->vdev_aux == NULL && vd == vd->vdev_top && !vd->vdev_ishole) {
-		vs->vs_fragmentation = vd->vdev_mg->mg_fragmentation;
-	}
-
+	int c, t, b;
 	/*
 	 * If we're getting stats on the root vdev, aggregate the I/O counts
 	 * over all top-level vdevs (i.e. the direct children of the root).
 	 */
-	if (vd == rvd) {
-		for (c = 0; c < rvd->vdev_children; c++) {
-			vdev_t *cvd = rvd->vdev_child[c];
+	if (!vd->vdev_ops->vdev_op_leaf) {
+		memset(vs->vs_ops, 0, sizeof (vs->vs_ops));
+		memset(vs->vs_bytes, 0, sizeof (vs->vs_bytes));
+		memset(vs->vs_disk_histo, 0, sizeof (vs->vs_disk_histo));
+		memset(vs->vs_total_histo, 0, sizeof (vs->vs_total_histo));
+		memset(vs->vs_queue_histo, 0, sizeof (vs->vs_queue_histo));
+		memset(vs->vs_queue_active, 0, sizeof (vs->vs_queue_active));
+
+		for (c = 0; c < vd->vdev_children; c++) {
+			vdev_t *cvd = vd->vdev_child[c];
 			vdev_stat_t *cvs = &cvd->vdev_stat;
 
+			__vdev_get_stats(cvd, cvs);
+
 			for (t = 0; t < ZIO_TYPES; t++) {
 				vs->vs_ops[t] += cvs->vs_ops[t];
 				vs->vs_bytes[t] += cvs->vs_bytes[t];
+				for (b = 0; b < VDEV_HISTO_BUCKETS; b++) {
+					vs->vs_disk_histo[t][b] +=
+						cvs->vs_disk_histo[t][b];
+					vs->vs_total_histo[t][b] +=
+						cvs->vs_total_histo[t][b];
+				}
+			}
+
+			for (t = 0; t < ZIO_PRIORITY_COUNT; t++) {
+				for (b = 0; b < VDEV_HISTO_BUCKETS; b++)
+					vs->vs_queue_histo[t][b] +=
+						cvs->vs_queue_histo[t][b];
+
+				vs->vs_queue_active[t] +=
+					cvs->vs_queue_active[t];
 			}
+
 			cvs->vs_scan_removing = cvd->vdev_removing;
 		}
+
+	} else {
+		/*
+		 * We're a leaf.  Just copy our ZIO active queue stats in.  The
+		 * other leaf stats are updated in vdev_stat_update().
+		 */
+		for (t = 0; t < ZIO_PRIORITY_COUNT; t++)
+			vs->vs_queue_active[t] =
+				vd->vdev_queue.vq_class[t].vqc_active;
 	}
+}
+
+void
+vdev_get_stats(vdev_t *vd, vdev_stat_t *vs) {
+	spa_t *spa = vd->vdev_spa;
```

Comment:
This is only used in the ASSERT below.  I'd suggest just replacing `spa` with `vd->vdev_spa` in the ASSERT.